### PR TITLE
gobject-introspection: fix non-meson cross builds

### PR DESCRIPTION
**Pull request by Artturin**
TODO write msg

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/gobject-introspection/wrappers/g-ir-scanner-qemuwrapper.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+# libdbusmenu-gtk3-aarch64-unknown-linux-gnu> /build/libdbusmenu-16.04.0/libdbusmenu-gtk/tmp-introspect2jhtiwwn/.libs/DbusmenuGtk3-0.4:
+# error while loading shared libraries: libdbusmenu-glib.so.4: cannot open shared object file: No such file or directory
+# in non-meson builds
+
+# see: https://github.com/void-linux/void-packages/blob/master/srcpkgs/gobject-introspection/files/g-ir-scanner-qemuwrapper
+# https://github.com/openembedded/openembedded-core/blob/c5a14f39a6717a99b510cb97aa2fb403d4b98d99/meta/recipes-gnome/gobject-introspection/gobject-introspection_1.72.0.bb#L74
+findfunc() {
+    while read -r d; do \
+        # some meson projects may have subprojects which use makefiles for docs(e.g. gi-docgen), ignore those as they will never be needed
+        if [[ -f "$d/Makefile" && "$d" != *"subproject"* ]]; then
```
Comment (Mindavi) [inline pkgs/development/libraries/gobject-introspection/wrappers/g-ir-scanner-qemuwrapper.sh:14]: What's with the *subproject*? I don't understand why it's matching on that here...

**Review by Mindavi — APPROVED** [review]
Further down the rabbit hole... This seems good other than one question I have.

Also, what did you build/try?

File: pkgs/development/libraries/gobject-introspection/wrappers/g-ir-scanner-qemuwrapper.sh (hunk 1 above)
Comment (Artturin) [reply pkgs/development/libraries/gobject-introspection/wrappers/g-ir-scanner-qemuwrapper.sh:14]: check the comment

**Comment by Artturin** [comment]
> Also, what did you build/try?

`pkgsCross.aarch64-multiplatform.libdbusmenu*` and `pkgsCross.aarch64-multiplatform.modemmanager`

this PR will also make the adding gobject-introspection to buildInputs PRs obsolete and the second commit will fix the todos in https://github.com/NixOS/nixpkgs/pull/184023

File: pkgs/development/libraries/gobject-introspection/wrapper.nix [hunk 2]
```diff
@@ -18,23 +18,38 @@ in
...
       export targetgir="${lib.getDev (targetPackages.gobject-introspection-unwrapped.override argsForTarget)}"
 
       substituteAll "${./wrappers/g-ir-compiler.sh}" "$dev/bin/g-ir-compiler"
       substituteAll "${./wrappers/g-ir-scanner.sh}" "$dev/bin/g-ir-scanner"
       substituteAll "${./wrappers/g-ir-scanner-lddwrapper.sh}" "$dev/bin/g-ir-scanner-lddwrapper"
+      substituteAll "${./wrappers/g-ir-scanner-qemuwrapper.sh}" "$dev/bin/g-ir-scanner-qemuwrapper"
       chmod +x "$dev/bin/g-ir-compiler"
       chmod +x "$dev/bin/g-ir-scanner"
       chmod +x "$dev/bin/g-ir-scanner-lddwrapper"
+      chmod +x "$dev/bin/g-ir-scanner-qemuwrapper"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/gobject-introspection/wrapper.nix:47]:
```suggestion
      chmod +x $dev/bin/g-ir-*
```

File: pkgs/development/libraries/gobject-introspection/wrappers/g-ir-scanner-qemuwrapper.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            GIR_EXTRA_LIBS_PATH="$(readlink -f "$d/.libs"):$GIR_EXTRA_LIBS_PATH"
+            export GIR_EXTRA_LIBS_PATH
+        fi
+    done < <(find "$NIX_BUILD_TOP" -type d)
+}
+
+findfunc
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/gobject-introspection/wrappers/g-ir-scanner-qemuwrapper.sh:21]: When this is only used once I don't think we need to wrap it in a function.

File: pkgs/development/libraries/graphene/default.nix [hunk 4]
```diff
@@ -59,6 +59,8 @@ stdenv.mkDerivation rec {
     gobject-introspection
     python3
     makeWrapper
+  ] ++ lib.optionals (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) [
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/graphene/default.nix:62]: We should really push this in a treewide change but thats a topic for another PR.

**Review by SuperSandro2000 — COMMENTED** [review]
Can we set strictDeps everywhere we fixed cross compile once?

Also please try to upstream those changes to make future updates easier.

File: pkgs/development/libraries/graphene/default.nix (hunk 4 above)
Comment (Artturin) [reply pkgs/development/libraries/graphene/default.nix:62]: `gtk-doc` always needs this when cross-compiling so it could propagate it

**Comment by Artturin** [comment]
will make a separate pr for the other fixes

**Pull request merged**